lib/ukallocstack/stack.c

```diff
@@ -0,0 +1,277 @@
+/* SPDX-License-Identifier: BSD-3-Clause */
+/* Copyright (c) 2024, Unikraft GmbH and The Unikraft Authors.
+ * Licensed under the BSD-3-Clause License (the "License").
+ * You may not use this file except in compliance with the License.
+ */
+
+#include <stddef.h>
+#include <string.h>
+#include <uk/alloc.h>
+#include <uk/alloc_impl.h>
+#include <uk/arch/ctx.h>
+#include <uk/arch/paging.h>
+#include <uk/assert.h>
+#include <uk/essentials.h>
+#if CONFIG_LIBUKVMEM
+#include <uk/vmem.h>
+#endif /* CONFIG_LIBUKVMEM */
+
+struct uk_allocstack {
+	struct uk_alloc a;
+	struct uk_alloc *og_a;
+#if CONFIG_LIBUKVMEM
+	__sz initial_size;
+	struct uk_vas *vas;
+#endif /* CONFIG_LIBUKVMEM */
+};
+
+static struct uk_allocstack *to_allocstack(struct uk_alloc *a)
+{
+	UK_ASSERT(a);
+
+	return __containerof(a, struct uk_allocstack, a);
+}
+
+static void *stack_malloc(struct uk_alloc *a, __sz size)
+{
+	__vaddr_t sp __maybe_unused;
+	struct uk_allocstack *sa;
+	int rc __maybe_unused;
+	__vaddr_t vaddr;
+	__sz real_size;
+
+	UK_ASSERT(a);
+
+	if (!size)
```

Comment:
Regarding this and the other `unlikely` comments. I meant to align with the default `memalign`/`posix-memalign` checks which don't have these. But now, looking back, it's better to have them here in this case, since this lib is only meant to be used by our own kernel internal allocators :+1: